### PR TITLE
[dy] Add prestart script support

### PR DESCRIPTION
**Pull request by dy46**
# Description

Support prestart script for workspaces. Users can write a custom method to edit the container config for their workspace before the container is started. The custom method takes in a container config as the argument, and returns an updated config:
```
def get_custom_configs(initial_container_config: Dict) -> dict:
    ...
```

In order to actually run the script in the workspace, Mage will

1. upload the script and the initial container config to a k8s ConfigMap
2. mount the script and config from the ConfigMap to an InitContainer
3. run the `pre-start.sh` bash script in the Init container
  1. create and run a `kubernetes-pre-start.py` python script based on the `get_custom_configs` method.

I'll update the documentation when this PR is closer to being ready to merge.

# How Has This Been Tested?

- [x] Tested locally on a kubernetes cluster

# Checklist
- [x] The PR is tagged with proper labels (bug, enhancement, feature, documentation)
- [x] I have performed a self-review of my own code
- [x] I have added unit tests that prove my fix is effective or that my feature works
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

cc: @wangxiaoyou1993

### CODE REVIEW

File: mage_ai/cluster_manager/kubernetes/workload_manager.py [hunk 1]
```diff
@@ -173,27 +176,80 @@ def create_workload(
...
-                **container_config,
-            }
-        ]
+                }
+            )
+
+            init_containers.append(
+                {
+                    'name': f'{name}-pre-start',
+                    'image': 'mageai/pre-start:latest',
```
Comment (wangxiaoyou1993) [inline mage_ai/cluster_manager/kubernetes/workload_manager.py:226]: Are you going to publish this docker image?

Comment (dy46) [reply mage_ai/cluster_manager/kubernetes/workload_manager.py:226]: yes i will once the PR is finalized

File: mage_ai/cluster_manager/kubernetes/workload_manager.py [hunk 2]
```diff
@@ -368,6 +425,7 @@ def add_service_to_ingress_paths(
     def delete_workload(self, name: str):
         self.apps_client.delete_namespaced_stateful_set(name, self.namespace)
         self.core_client.delete_namespaced_service(f'{name}-service', self.namespace)
+        self.core_client.delete_namespaced_config_map(f'{name}-pre-start', self.namespace)
```
[the remark this answers is not in the document]
Comment (dy46) [reply mage_ai/cluster_manager/kubernetes/workload_manager.py]: good point, i updated it

File: scripts/pre-start/kubernetes_pre_start_template.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import argparse
+import json
+from typing import Dict
+
+from kubernetes import client, config
+
+{{ config_code }}
```
Comment (wangxiaoyou1993) [inline scripts/pre-start/kubernetes_pre_start_template.py:7]: Have you tried testing with an invalid script? What does the error look like? Is it hard to debug?

Comment (dy46) [reply scripts/pre-start/kubernetes_pre_start_template.py:7]: you should be able to see errors in the init container logs if it fails. we could potentially show those error logs in the UI

**Comment by dy46** [comment]
@wangxiaoyou1993 added validation

**Review by wangxiaoyou1993 — APPROVED** [review]

**Pull request merged**